dis_snek/api/http/http_client.py

```diff
@@ -118,98 +158,71 @@ async def request(
-                    pass
-                raise
-            except (Forbidden, NotFound, DiscordError, HTTPException):
-                raise
-            except Exception as e:
-                try:
-                    lock.release()
-                except RuntimeError:
-                    pass
-                log.error("".join(traceback.format_exception(type(e), e, e.__traceback__)))
-                break
-        if lock.locked():
-            # be clean and make sure we unlock
-            lock.release()
+                    if self.__session.closed:
+                        await self.login(self.token)
+
+                    await self.global_lock.rate_limit()
+
+                    async with self.__session.request(route.method, route.url, **kwargs) as response:
+                        result = await response_decode(response)
+                        r_limit_data = self._parse_ratelimit(response.headers)
+
+                        if response.status == 429:
+                            # ratelimit exceeded
+                            log.error(
+                                f"{route.endpoint} Has exceeded it's ratelimit! Reset in {r_limit_data['delta']} seconds"
+                            )
+                            await lock.deferred_unlock(r_limit_data["delta"])
+                            continue
+                        elif r_limit_data["remaining"] == 0:
+                            # Last call available in the bucket, lock until reset
+                            log.debug(
+                                f"{route.endpoint} Has exhausted its ratelimit! Locking route for {r_limit_data['delta']}"
+                            )
+                            await lock.blind_deferred_unlock(r_limit_data["delta"])
+
+                        elif response.status in {500, 502, 504}:
+                            # Server issues, retry
+                            log.warning(
+                                f"{route.endpoint} Received {response.status}... retrying in {1 + attempt * 2} seconds"
+                            )
+                            await asyncio.sleep(1 + attempt * 2)
+                            continue
+
+                        if not 300 > response.status >= 200:
+                            await self._raise_exception(response, route, result)
+
+                        log.debug(f"{route.endpoint} Received {response.status}, releasing")
+                        return result
+                except OSError as e:
+                    if attempt < self._max_attempts - 1 and e.errno in (54, 10054):
+                        await asyncio.sleep(1 + attempt * 2)
+                        continue
+                    raise
+                except (Forbidden, NotFound, DiscordError, HTTPException):
+                    raise
+                except Exception as e:
+                    log.error("".join(traceback.format_exception(type(e), e, e.__traceback__)))
+                    break
```

Comment:
What's wrong with just letting it propogate?

Reply:
Valid point, i believe this was left over from the previous smelly code. I'll remove this except